### DISCUSSION TITLE
General Overview

### DISCUSSION THREAD

**Discussion by afonsolage** [afonsolage/wabi#1] [Ideas]
## Introduction
In this discussion, I'll write some thoughts on how to create an interaction between app and module. 

Note that those thoughts may not be connected to each order in the order they appear, but I'll try my best to keep short and organized.

Note also that I assume that you already know Bevy and it's terminology.

## Definitions
- *_App_* is where the Bevy `App` is built and ran;
- *_Mod_* is the wasm module which is compiled, loaded and executed by _App_;
- *_Host_* is the wasm runtime. On native builds, it refers to a wasm runtime, like `wasmtime` or `wasmer`. On web builds, it refers to JavaScript runtime of the browser, like `V8` or `SpiderMonkey`;
- *_Guest_* is any wasm module which is executed by _Host_. When running native builds, it refers to _Mod_. When running on web builds, it may refer to _App_ or _Mod_, depending of the context;
- *_API_* is the protocol defined to allow the communication between _App_ and _Mod_;
- *_Mod Asset_* is the raw binary data of a compiled wasm module;
- *_Mod Instance_* is an executable instance of a _Mod Asset_;
- *_Mod System_* a system exported from a _Mod Instance_;
- *_Mod Run Criteria_* is a [run criteria](https://bevy-cheatbook.github.io/programming/run-criteria.html) which may be informed by any *_Mod System_* that must be evaluated to true in order for the _Mod System_ to be ran;
- *_Contex_* is the data required by any _Mod System_ to run;
- *_Context Switching_* means switch execution from `Host` to `Guest` or _vice-versa_, either by calling a function or returning from a previous function call. This is an analogy with [Context Switch](https://en.wikipedia.org/wiki/Context_switch) of threads.

## Native and Web targets
One of the biggest challenges of this project is to enable seamless modding in either native or web (and in the future, mobile) platforms. The main reason why it is challenging, is because when targeting web (`wasm32-unknown-unkown`) the _App_ is a _Guest_, since it's running inside the browser, and the JS runtime of the browser is the _Host_. On the other hand, when running on native platform, the _App_ is also the _Host_ and the _Mod_ is also the _Guest_.

To summarize, always keep in mind that:
- Native: _Host App_ <-> _Guest Mod_;
- Web: _Guest App_ <-> _Host Browser_ <-> _Guest Mod_;

## App and Mod interaction overview
_App_ will use `exclusive_systems` to run _Mod System_, since it's not possible at moment to add or remove dynamic systems.

1. _App_ loads and compile¹ _Mod Asset_;
2. _App_ instantiate a new _Mod Instance_;
3. _App_ calls `__wabi_alloc` on _Mod Instance_ to allocate it's memory buffer;
4. _App_ calls `__wabi_init` on _Mod Instance_ to initialize any internal state;
5. _App_ calls `__wabi_systems` on _Mod Instance_ to get all systems which will be run;
6. On each `Stage`, _App_:
    1. Checks for any exported _Mod System_ to be ran on current stage;
    2. Prepares a _Context_ for each _Mod System_;
    3. Run all _Mod System_ with appropriated _Context_;
    4. Apply any modifications made on _Context_ to the `World`;

¹ <sub>Even tho the wasm is already compiled, it is compiled to a bytecode format. Before running it, the _Host_ needs to do some platform specific optimizations.</sub>

## App and Mod data sharing
Since _Mod_ runs in a sand boxed environment, it's not possible to share memory from _Host_ to _Guest_. In order to share data, a [WebAssembly.Memory](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory) must be used.

When a _Mod Instance_ is instantiated, _App_ asks _Mod_ to allocate a fixed size buffer and returns a pointer for this allocated buffer. This buffer will be allocated for the entire lifetime of the _Mod_, to avoid memory corruption. 

Since _App_ has control over which `WebAssembly.Memory` was given to each _Mod Instance_, when _App_ needs to share some data with _Mod_, it writes the data on `WebAssembly.Memory` buffer offsetted by the returned pointer address, which is relative to the memory buffer. After writing, _App_ asks for _Mod_ to read the buffer, passing the length of the written data. That way, _Mod_ can safely read the data.

When _Mod_ needs to share some data with _App_, a similar approach is used, whereas _Mod_ writes the data on allocated buffer and than send the written length to the _App_, which in turns reads from the memory buffer, using the previous returned pointer offset and the given length.

Note that due to this workflow, the _Mod_ should not relay on the memory buffer to run any logic, since a simple log call may overwrite the buffer content, instead, before running any function call by the _App_, the _Mod_ must make use of the data, making a copy of the relevant data.

## Mod definition
Each _Mod_ must export at least those functions, in order to interact with _App_:
- `__wabi_alloc` which will allocate an internal buffer to share data between _App_ and _Mod_. This will be called before any other _Mod_ function;
- `__wabi_init` which will initialize any internal state needed by the module. This will be called only once, after `__wabi_alloc_`;
- `__wabi_systems` which will describe what systems should be run. This will be called at the beginning of each Bevy event loop, to allow systems being enabled or disabled;

Each _Mod_ must import the following _Host_ functions:
- `__wabi_send_action` which will request the _App_ to execute some action;

## Mod Run Criteria
Since there is some overhead involved when _Context Switching_, it is mandatory for the _App_ to only run _Mod System_ when it is absolutely needed. To do so, a _Mod System_ can inform one or more _Mod Run Criteria_ s that must evaluate to true in order for the system to be ran.

These are the initial planned _Mod Run Criteria_s:
- `ComponentAdded`
- `ComponentChanged`
- `ComponentRemoved`
- `ResourceAdded`
- `ResourceChanged`
- `ResourceRemoved`
- `QueryNotEmpty`
- `EventNotEmpty`

## Mod System definition
Each Mod must export at least one _Mod System_, which will be ran when the all _Mod Run Criteria_s are met, if specified. In order to export a _Mod System_ a _Mod_ must inform one or more:
 - Exported system function name;
 - Required _Context_;
 - Zero or more _Mod Run Criteria_s;

The exported system function must accept one argument, which is the size of the supplied _Context_ when the function will be ran.

## Context
The _Context_ used by _Mod System_ will contains zero or more of the following Bevy types:
- [`Entity`](https://bevy-cheatbook.github.io/programming/ec.html#entities);
- [`Component`](https://bevy-cheatbook.github.io/programming/ec.html#components);
- [`Resource`](https://bevy-cheatbook.github.io/programming/res.html#resources);
- [`Commands`](https://bevy-cheatbook.github.io/programming/commands.html#commands);
- [`EventReader`](https://bevy-cheatbook.github.io/programming/events.html#events);
- [`EventWriter`](https://bevy-cheatbook.github.io/programming/events.html#events);

Types like `Component` and `Resource` will have a internal bit flag which will indicate if the component was updated or not, this will be used to avoid triggering [change detection](https://bevy-cheatbook.github.io/programming/change-detection.html) when not needed.

After each _Mod System_ execution, the _Context_ will be read by _App_ to check for modifications, as follows:
- `Component`s, if the changed bit flag is set;
- `Resource`s, if the changed bit flag is set;
- `EventWriter` will have all it's items dispatched;
- `Commands` will have all it's commands flushed;

## Context Switching
Since every _Mod_ should be ran in a single thread, the _Context Switching_ is quite easy to handle, since it'll follow a simple call-n-return function:
1. _App_ calls the _Mod System_ exported function supplying the required _Context_;
    1. _Mod System_ starts to execute the logic based on given _Context_;
    2. If needed, _Mod System_ calls `__wabi_send_action` and wait for return;
    3. _App_ runs the callback function associated with `__wabi_send_action` and return a value, if needed;
    4. _Mod System_ check the returned value and continue execution;
    5. Repeat 2, 3 and 4 until  _Mod System_ finishes it's execution;
2. _App_ continues the execution;

Also there is a little overhead when switching context between _App_ and _Mod_. Some initial benchmarks lead to the following results on release mode:
### Function call
- (native) _Host App_ to _Guest Mod_: *~1μs*
- (web) _Guest App_ to _Host_ to _Guest Mod_: *~2μs*
### Share state (12k buffer size reading)
- (native) _Host App_ to _Guest Mod_: *~2μs*
- (web) _Guest App_ to _Host_ to _Guest Mod_: *~4μs*

So we can rough estimate that a _Context Switching_ where _Mod System_ send some data to _App_ and it in turns returns back some data, will have a *~6μs* overhead on native and *~12μs* on web.

Note that those are very rough estimations, and depending on the machine, platform, browser and data shared, may change A LOT, but for the purpose of this document, it will help to make some initial decisions.

**Comment by afonsolage** [comment]
**Edit: RPC is out-of-scope for now. See https://github.com/afonsolage/wabi/issues/2**

I'll keep the bellow text just for future reference.

---

# API RPC
Another important aspect of any API for modding is the ability to forward calls from _Mod_ to _App_, so _Mod_ doesn't have to reimplement all the logic which already exists in _App_.

When dealing with wasm, the API RPC must be language agnostic, which means any language which is able to target wasm should be able to use the API RPC without any problems.

To understand the challenge here, take for instance an example:
```rust
let transform = Transform::from_xyz(10.0, 11.0, 12.0)
    .looking_at(Vec3::ZERO, Vec3::Y)
    .with_scale(Vec3::splat(1.0));
```
Since _App_ and _Mod_ doesn't share the same stack, one must use the memory buffer to transfer data between _App_ and _Mod_ as explained before. Breaking down the above sample code, this is what it will look like in the _Guest_:

1. `Transform::from_xyz`:
    1. _Guest_ writes to the memory buffer a list of floats;
    2. _Guest_ calls `__wabi_send_action` passing the written buffer length and an action code asking to create a new `Transform` from xyz;
    3. _App_ reads the memory buffer with the given length;
    4. _App_ dispatches the action code to a `TransformProxy` which will translate to the actual `Transform::from_xyz`;
    5. _App_ gets the new crated `Transform`, serializes it and write to memory buffer, returning the length written on buffer;
    6. _Guest_ reads the memory buffer and deserializes the `Transform`;

2. `.looking_at(Vec3::ZERO, Vec3::Y)`
    1. _Guest_ serializes the `Transform`, `Vec3::ZERO` and `Vec3::Y`;
    2. _Guest_ writes the serialized data to memory buffer; 
    3. _Guest_ calls `__wabi_send_action` passing the written buffer length and an action code relative to `Transform::looking_at`;
    4. _App_ reads the memory buffer with the given length;
    5. _App_ dispatches the action code to a `TransformProxy` which will translate to the actual `Transform::looking_at`;
    6. _App_ gets the new returned `Transform`, serializes it and write to memory buffer, returning the length written on buffer;
    8. _Guest_ reads the memory buffer and deserializes the `Transform`;
  
3. `Vec3::splat(1.0)`
    1. _Guest_ calls a internal function with will create a new `Vec3` using the `splat` function;
    2. _Guest_ writes to the memory buffer a float;
    3. _Guest_ calls `__wabi_send_action` passing the written buffer length and an action code relative to `Vec3::splat`;
    4. _App_ reads the memory buffer with the given length;
    5. _App_ dispatches the action code to a `Vec3Proxy` which will translate to the actual `Vec3::splat`;
    6. _App_ gets the new returned `Vec3`, serializes it and write to memory buffer, returning the length written on buffer;
    7. _Guest_ reads the memory buffer and deserializes the `Vec3`;
    8. _Guest_ returns the new created `Vec3`;

4. `.with_scale(Vec3::splat(1.0))`
    1. _Guest_ serializes the `Transform` and the returned `Vec3`;
    2. _Guest_ writes the serialized data to memory buffer; 
    3. _Guest_ calls `__wabi_send_action` passing the written buffer length and an action code relative to `Transform::with_scale`;
    4. _App_ reads the memory buffer with the given length;
    5. _App_ dispatches the action code to a `TransformProxy` which will translate to the actual `Transform::with_scale`;
    6. _App_ gets the new returned `Transform`, serializes it and write to memory buffer, returning the length written on buffer;
    7. _Guest_ reads the memory buffer and deserializes the `Transform`;

That's a lot of _Contex Switching_ and memory read/writes, but there are a lot of optimizations that can be done also, for instance:
1. Cache objects when applying chained `self` functions;
2. Use zero-copy ser/de to avoid copying data around;
3. Batch actions, so the chained transformations happens only once on _App_;
4. Use multiple memory buffers to reduce the amount of copy;

But those optimizations is out of scope of this document, since it's very risk to assume the benefits of those optimizations without proper testing and benchmarking, which will be done after there is a MVP working.

## Proxy types
There should be two kinds of proxies:
- `ModProxy` which will forward the call to the _App_, providing a proper function code to map to a `AppProxy`;
- `AppProxy` which will receive a function code and forward for the actual underlying type function;

### ModProxy
A `ModProxy` is a mapping between functions id and functions parameters and results. When a method is called, `ModProxy` will expect two types of parameters:
 - A primitive type;
 - Another `ModProxy` type;
 
Then it'll serialize the parameters and forward the call to `__wabi_proxy_call` passing the serialized data length, `ProxyTypeId` and ProxyFunctionId`.

### AppProxy
Each `AppProxy` will have a underlying type `T: Reflect`, a `ProxyTypeId` and for each public function, a `ProxyFunctionId`, which is unique within an `AppProxy`. Both `ProxyTypeId` and `ProxyFunctionId` are `u32` since it can be passed as function argument, avoiding using the memory buffer when requesting a RPC call;

When a dispatch action is request, `AppProxy` will deserialize the given `Reflect` dynamic object, mapping to the correct underlying function arguments and then call it, If the underlying function uses `&mut self`, serialize the underlying type and the returned value; if any, otherwise, only serialize the returned value.

**Comment by linonetwo** [comment]
How does this solution different from extism's solution, can this be built on extism?
